src/cpp/src/tokenizer.cpp

```diff
@@ -374,39 +387,28 @@ class Tokenizer::TokenizerImpl {
                         "Chat template wasn't found. This may indicate that the model wasn't trained for chat scenario."
                         " Please add 'chat_template' to tokenizer_config.json to use the model in chat scenario."
                         " For more information see the section Troubleshooting in README.md");
-
-        // Jinja2Cpp does not support Python-style slicing, e.g. [1:].
-        // If chat template contains such slicing, we replace it with custom function `slice()` (user-defined callable) 
-        // that is defined below and does the same list slicing logic.
-        std::string slice_string = "messages[1:]";
-        std::string replacement_slice_string = "slice(messages, 1)";
-        size_t slice_pos = chat_tpl.find(slice_string);
-        if (slice_pos != std::string::npos) {
-            chat_tpl.replace(slice_pos, slice_string.length(), replacement_slice_string);
-        }
+        jinja2::TemplateEnv env;
+        env.GetSettings().lstripBlocks = true;
+        env.GetSettings().trimBlocks = true;
+        jinja2::Template tpl(&env);
+        tpl.Load(chat_tpl);
+        
         jinja2::UserCallable slice_callable = jinja2::MakeCallable(
-            [](const jinja2::ValuesList& list, const int64_t start) {
-                if (list.empty())
-                    return jinja2::Value();
+            [](const jinja2::GenericList& messages, const size_t& start) {
                 jinja2::ValuesList result;
-                int64_t stop = list.size();
-                int64_t step = 1;
-                for (int64_t i = start; i < stop && i < list.size(); i += step)
-                {
-                    result.push_back(list.at(i));
+
+                size_t iter_num = 0;
+                for (auto message = messages.begin(); message != messages.end(); message++, iter_num++) {
+                    if (iter_num < start)
+                        continue;
```

Comment:
You can probably initialize `message` iterator with `messages.begin() + start`.

Even better. std::vector allows construction from iterators. Maybe `jinja2::ValuesList` can do that too. If that's the cause it all collapses to `return jinja2::ValuesList{messages.begin() + start, messages.end());`

Reply:
>You can probably initialize message iterator with messages.begin() + start.
I tried, it didn't work, i tried also `messages.GetAccessor()` still no luck. So this is the simples solution i came up with

Reply:
Does
>initialize `message` iterator with `messages.begin() + start`

work?